HPCC-13030 Add row tracing support to engines

shamser opened the pull request:
Return last 10 rows for a given activity & edge, when xml submitted to port 9877 (or as configured in global) in the form:

  <debug source="ACTIVITY_ID" edge="EDGE_ID"></debug>

richardkchapman committed:
HPCC-13030 Add row tracing support to engines

Implement row tracing for Thor

shamser commented:
@richardkchapman @jakesmith Not complete but would appreciate interim comments. Thanks.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on thorlcr/activities/choosesets/thchoosesetsslave.cpp:
```diff
@@ -250,6 +250,7 @@ class InputCounter : public CSimpleInterface, implements IThorDataLink
     virtual CActivityBase *queryFromActivity();
     virtual void dataLinkSerialize(MemoryBuffer &mb);
     unsigned __int64 queryTotalCycles() const;
+    virtual void debugRequest(MemoryBuffer &msg);
```
minor: would be good to make the other IThorDataLink interface definitions here, declared with virtual as well at this juncture.

richardkchapman replied:
Why?

jakesmith replied:
Consistency.
All are virtuals and declared virtuals here, apart from queryTotalCycles(),.

jakesmith commented on thorlcr/activities/thactivityutil.ipp:
```diff
@@ -165,6 +165,7 @@ public:
     {
         return icount; 
     } 
+    virtual void debugRequest(MemoryBuffer &msg) {}
```
v.trivial: normally have space between brackets.

jakesmith commented on thorlcr/graph/thgraph.hpp:
```diff
@@ -847,6 +848,7 @@ class graph_decl CJobBase : public CInterface, implements IDiskUsage, implements
     unsigned __int64 getOptUInt64(const char *opt, unsigned __int64 dft=0) { return (unsigned __int64)getOptInt64(opt, dft); }
 
     virtual void abort(IException *e);
+    virtual void debugRequest(CMessageBuffer &msg, const char *request) const {}
```
v.trivial: normally have space between brackets.

richardkchapman replied:
I don't...

jakesmith commented on thorlcr/graph/thgraphslave.hpp:
```diff
@@ -64,8 +65,8 @@ class graphslave_decl CSlaveActivity : public CActivityBase
-    void appendOutput(IThorDataLink *itdl) { outputs.append(itdl); };
-    void appendOutputLinked(IThorDataLink *itdl) { itdl->Link(); appendOutput(itdl); };
+    void appendOutput(IThorDataLink *itdl) { outputs.append(itdl);};
+    void appendOutputLinked(IThorDataLink *itdl){ itdl->Link(); outputs.append(itdl); };
```
trivial: spacing before {

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,145 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+                        DBGLOG("Reading debug command from socket...");
+                        if (!ssock.readBlock(rawText, WAIT_FOREVER, NULL, continuationNeeded, isStatus, 1024*1024))
+                        {
+                            if (traceLevel > 8)
```
Where's 'traceLevel' coming from? Hm, looks like external of roxiehelper.hpp.
Shouldn't use that.
Thor doesn't really have a traceLevel, though it should, but separate issue.

For now I'd just remove this condition and always trace this message.

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,145 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+                        {
+                            if (traceLevel > 8)
+                            {
+                                StringBuffer b;
```
minor: unused, should be removed.

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,145 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+                            if (traceLevel > 8)
+                            {
+                                StringBuffer b;
+                                DBGLOG("No data reading query from socket");
```
WARNLOG may be more appropriate.

jakesmith commented on thorlcr/slave/slavmain.cpp:
```diff
@@ -511,6 +511,25 @@ class CJobListener : public CSimpleInterface
+                    {
+                        StringAttr jobKey;
+                        msg.read(jobKey);
+                        PROGLOG("DebugRequest: %s", jobKey.get());
```
minor: I'd combine this line of tracing together with the one in the if (job) block below.

jakesmith commented on thorlcr/slave/slwatchdog.cpp:
```diff
@@ -132,6 +132,29 @@ class CGraphProgressHandlerBase : public CSimpleInterface, implements ISlaveWatc
+        {
+            // MORE - this code should be in CGraphBase
+            CGraphBase &graph = activeGraphs.item(g);
+            CGraphElementBase *element = graph.queryElement(actId);
```
I think the above 'MORE - this code should be in CGraphBase' is still true, would be better if moved into CSlaveGraph

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+#include "limits.h"
+
+
+class CTracingThorDataLink :  implements CInterfaceOf<IThorDataLink>
```
trivial: spurious double space

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+            // NOTE - we don't output every item in the list to be 
+            // consistent with the other dump() implementation
+            // Note - whilst dumping, does not clear ring buffer
+            // so that it is available later
```
The rowLogCache, is only ever in use by 1 thread, so doesn't need to be +1 in size and skip the rowBuffer[rowHeadp]'th element.
I think this should be sized to N (e.g. default 10) and return all.

It will make it clearer that different and not in conflict like the other buffers,

shamser replied:
TODO

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+                if (!row) continue;
+
+                CommonXmlWriter xmlwrite(XWFnoindent|XWFtrim);
+                helper->queryOutputMeta()->toXML((const byte*)(row.get()), xmlwrite);
```
Should we be serializing the row data here and converting to XML on the master?
Not sure there's a lot of point...
@ghalliday

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+        void dump(MemoryBuffer &mb, TraceQueue &rowLogCache, IHThorArg *helper)
+        {
+            // NOTE - we don't output every item in the list as one
+            // COULD have been updated after we swapped the buffers
```
> we don't output every item

Comment could do with clarifying..
I think it's that you start at next head position (rowHeadp+1) and cycle through traceRowBuffersize-1 rows.
Intentionally skipping 1st (one that was at rowHeadp when started), because it's possible the writer could have been setting/releasing the row at that position.

jakesmith replied:
Also (and maybe we discussed this(?), it does mean that if there's that race condition, that rows will appear out of sequence. e.g.:
+ buffers[0] = [a],[b],[*c] (c about to be written)
+ dump swaps, a,b added to tracing, buffers[1] now active
+ buffers[1] = [d],[e],[*f] (f about to be written)
+ dump swap, d,e added to tracing, buffers[0] now active
+ buffers[0] = [],[],[c],[g],[h],[*i] (i about to be written)
+ dump swap, c,g,h added to tracing, buffers[1] now active
+ buffers[1] = [],[],[f],[j],[k],[*l] (l about to be written)
+ dump swap, f,j,k added to tracing..


Should we just accept that, and just put a comment to mention that this can happen?

Or could we look at the other buffer (the one about to be swapped in) for the trailing row and add to the tracing circular ring, before swapping in ?
I think that could mean there's still a window if 2 swaps happened very quickly, where it would still be active and not picked up and therefore still lead to out order rows, but in practice with the time taken between 2 swaps over 2 debugRequest() hops that isn't going to happen.

shamser replied:
Good point but not sure how fixing this yet. Queueing the non-active buffer before swapping in the Active buffer will cause other issues. We do not know that the items in the non-active buffer are *immediately* before the ones in the active buffer - the active buffer may have cycled a number of times so that there is a gap in the sequence between the active and non-active buffers.

The swaps do not necessarily have to happen in a short space of time.  If the rows are being generated slowly and, so, passing through the proxy, at a slow rate (such as one row every minute), then it's possibly even requests over several minutes the rows could be out of order.  (Of course, if the rows are arriving slowly, then the probability that the request is made at the exact time that a row is being enqueued is small).

jakesmith replied:
>We do not know that the items in the non-active buffer are immediately before the ones in the active buffer - the active buffer may have cycled a number of times so that there is a gap in the sequence between the active and non-active buffers.

That's true, but if that's true, the non-active buffer that is about to be swapped out and added to the tracingQueue will be full, so that the extra row added (that as you point out may be out of date / old, with many since) will be roll off as the new swapped out buffer rows are added to the tracing queue.

It's only if there active queue being swapped out is partially full that it really matters. 

I think this applies what ever the rate.

shamser replied:
@jakesmith Ok.. got it.. will change on next commit.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+                if (++rowPos==traceRowBufferSize) rowPos = 0; // NOTE - skipping first item intentional
+                OwnedConstThorRow row = rowBuffer[rowPos].getClear();
+
+                if (row) rowLogCache.enqueue(row);
```
The buffers, will have end-of-group (NULL) entries.
What did we say we'd do about end-of-group markers for tracing?
Ignore them? Add then trace them, in that case the above code would only want to skip forward if double NULL, meaning that the circular buffer was truly clear and they weren't eog.
@richardkchapman @ghalliday

shamser replied:
Double null does not necessarily mean the end of the list has been reached.  See below where 3 rows have been added:
Buffer:  [row1]  [row2] [row3] [  ] [  ] [  ] [  ] [  ] [  ] [  ] [  ]
*[rown] = has data
*[  ] = empty/no data
* headp would be pointing to 4th item (empty item)

In our code, we start at headp+1, which means there will be 7 *empty* items before the first item would be reached.

jakesmith replied:
Yes, that's what I meant in comment above re. skipping
> ..skip forward if double NULL, meaning that the circular buffer was truly clear and they weren't eog.

richardkchapman replied:
I thought we were not adding null rows (in which case all this discussion is moot)

jakesmith replied:
It isn't at the moment, as you say if we happy not to, then the discussion is moot.
That was my question (above).
```
What did we say we'd do about end-of-group markers for tracing?
Ignore them? Add then trace them, ..
```

It's not an efficiency issue, but complicates the handling a bit.
I don't think they are very important..

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+    class TraceQueue rowBufferLogCache;
+    atomic_t rowBufInUse;
+
+    static const unsigned traceRowBufferSize=11;  // we only output 10
```
Would be worth making configurable option passed into TraceQueue, via a getOptBool() fetched from job.

shamser replied:
@jakesmith What option do you mean we should get via getOptBool?

jakesmith replied:
whoops didn't mean Bool.
I meant for 'traceRowBufferSize', e.g. acitivty.getPropInt("numTraceRows", 10); could be passed into CTracingThorDataLink constructor.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+    }    // can only be called on stepping fields.
+    virtual IInputSteppingMeta *querySteppingMeta() { return thorDataLink->querySteppingMeta(); }
+    virtual bool gatherConjunctions(ISteppedConjunctionCollector & collector) { return thorDataLink->gatherConjunctions(collector); }
+    virtual void resetEOF() { thorDataLink->resetEOF();}
```
trivial: space before }

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+    virtual bool gatherConjunctions(ISteppedConjunctionCollector & collector) { return thorDataLink->gatherConjunctions(collector); }
+    virtual void resetEOF() { thorDataLink->resetEOF();}
+
+    virtual void getMetaInfo(ThorDataLinkMetaInfo &info) { thorDataLink->getMetaInfo(info);}
```
trivial: space before }

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+    virtual void resetEOF() { thorDataLink->resetEOF();}
+
+    virtual void getMetaInfo(ThorDataLinkMetaInfo &info) { thorDataLink->getMetaInfo(info);}
+    virtual CActivityBase *queryFromActivity() {return thorDataLink->queryFromActivity();} // activity that has this as an output
```
trivial: spacing after { , before }

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+
+    virtual void getMetaInfo(ThorDataLinkMetaInfo &info) { thorDataLink->getMetaInfo(info);}
+    virtual CActivityBase *queryFromActivity() {return thorDataLink->queryFromActivity();} // activity that has this as an output
+    virtual void dataLinkSerialize(MemoryBuffer &mb) {thorDataLink->dataLinkSerialize(mb);}
```
trivial: spacing after { , before }

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+    virtual void getMetaInfo(ThorDataLinkMetaInfo &info) { thorDataLink->getMetaInfo(info);}
+    virtual CActivityBase *queryFromActivity() {return thorDataLink->queryFromActivity();} // activity that has this as an output
+    virtual void dataLinkSerialize(MemoryBuffer &mb) {thorDataLink->dataLinkSerialize(mb);}
+    virtual unsigned __int64 queryTotalCycles() const { return thorDataLink->queryTotalCycles();}
```
trivial: spacing before }

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+        enqueueRowForTrace(row);
+        return row;
+    }
+    virtual void stop() {
```
style: { on new line

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+        thorDataLink->stop();
+    }                              // after stop called NULL is returned
+
+    inline const void *ungroupedNextRow() { return thorDataLink->ungroupedNextRow();}
```
trivial: spacing before }

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,163 @@
+    inline const void *ungroupedNextRow() { return thorDataLink->ungroupedNextRow();}
+
+    CTracingThorDataLink(IThorDataLink *_input, IHThorArg *_helper)
+    : thorDataLink(_input), helper(_helper)
```
style: normally indent here by 4 spaces

jakesmith commented:
@shamser - looks good, a few comments, mostly minor/trivial.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

shamser commented:
@jakesmith @richardkchapman Code modified as per code review comments and few otherimprovements.  Thanks.

richardkchapman commented:
Re tracking of NULL rows/group boundaries - I don't think it's a particularly significant feature and certainly would be inclined to keep the code simple. At some point (post 6.0) if people like this tracing we may want to start enhancing it with things like timestamps on the rows and we can add a note of group boundaries at the same time if we want. there's no reason we have to mark group boundaries via NULLs saved in the rowstream just because that's now they appear on the interface...

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

shamser commented:
@richardkchapman @jakesmith There are couple of issues that I haven't been able to resolve yet
1) I haven't been successful in replacing the IThorDataLink in the outputs.  I have modified appendOutput and appendOutputLinked in CSlaveActivity to replace IThorDataLink with CTracingThorDataLink.  Unfortunately, that only cause the thor slave to crash - i'm still looking into this.
2) For one activity (NSpliterSlaveActivity), I have haven't been able to write out the row using helper->queryOutputMeta->ToXml().  The helper->queryOutputMeta->getRecordSize() returns a very large number: 3 billion+.  The same code is able to return the rows for all other activities.  (CTracingThorDataLink::dump())

richardkchapman commented:
If you have an example of the ECL you used for the NSplitterSlaveActivity issue I can take a look

GordonSmith commented on esp/src/CMakeLists.txt:
```diff
@@ -16,7 +16,7 @@
 project ( eclwatch ) 
 
-set ( ECLWATCH_BUILD_STRATEGY "" CACHE STRING "Override ECL Watch Build Strategy ('SKIP', 'SOURCES', 'IF_MISSING', 'INCLUDE_MAP_FILES')" )
+set ( ECLWATCH_BUILD_STRATEGY "IF_MISSING" CACHE STRING "Override ECL Watch Build Strategy ('SKIP', 'SOURCES', 'IF_MISSING', 'INCLUDE_MAP_FILES')" )
```
This is not a good option as the default.

shamser replied:
I'll change this back.

shamser commented:
@richardkchapman @jakesmith @GordonSmith The Thor debugger is working end-to-end (from ECLEDI to Thor).  No known issues remain.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on thorlcr/slave/slwatchdog.cpp:
```diff
@@ -132,6 +132,35 @@ class CGraphProgressHandlerBase : public CSimpleInterface, implements ISlaveWatc
+        StringBuffer edgeString, sActId;
+        req->getProp("@edgeId", edgeString);
+        // Make sure it has sensible values
+        if (!edgeString.length() || edgeString.length()>99 ) return;
```
>99 a little arbitrary
I supposed strictly speaking limit is max digit length of 2 unsigned 32bit numbers + '_' 

But I'd probably just validate that when you perform the atoi's and make sure actId is not 0 too.

jakesmith replied:
extra space before closing bracket

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+                    }
+                    catch (...)
+                    {
+                        if (recSize>9999)
```
Not _that_ unusual to have record size bigger than that size (especially where child datasets are concerned).
And if wasn't, doesn't trace anything?

Not sure what the logic is here.

jakesmith commented on cmake_modules/FindJNI.cmake:
```diff
@@ -195,6 +195,7 @@ JAVA_APPEND_LIBRARY_DIRECTORIES(JAVA_AWT_EXPANDED_INCLUDE_DIRECTORIES
   /usr/lib/jvm/java-6-sun-1.6.0.00/include       # can this one be removed according to #8821 ? Alex
   /usr/lib/jvm/java-7-openjdk-{libarch}/include  # Ubuntu 13.04 location
   /usr/lib/jvm/java-6-openjdk-{libarch}/jre/lib/{libarch}  # Ubuntu 12.10 location
+  /usr/lib/jvm/java-6-openjdk-{libarch}/include
```
This doesn't seem related to this commit

shamser replied:
It isn't related to this change.  Would you  normally create a new ticket for single line changes like this?

richardkchapman replied:
Yes

jakesmith commented on thorlcr/activities/choosesets/thchoosesetsslave.cpp:
```diff
@@ -249,8 +249,9 @@ class InputCounter : public CSimpleInterface, implements IThorDataLink
-
+    virtual unsigned __int64 queryTotalCycles() const;
+    virtual void debugRequest(MemoryBuffer &msg);
+    virtual void setInputSplitterOutputBase(CSplitterOutputBase *_input, unsigned _id=0){}
```
trivial: missing space before {} (inconsistent with elsewhere)

But I'm not sure what this method is for.. The fact it is CSplitterOutputBase, but in all IThorDataLink definitions sounds wrong.

shamser replied:
In NSplitterSlaveActivity, please have a look at the definition of ensureInputsConfigured(). 

Original code had this:

CDelayedInput *delayedInput = (CDelayedInput *)outputs.item(o2);

'outputs' items were IThorDataLink * but all items  in this particular 'outputs' array were created by this acitivity and it was known that these IThorDataLink * from outputs could be up cast to CDelayedInput.  This code just assumed all IThorDataLink * items in 'outputs' could be upcasted to  CDelayedInput *

When the tracing code was enabled, the assumption was no longer true.  When CDelayedInput * is added to 'outputs', if tracing is enabled, the CTracingThorDataLink (wrapping CDelayedInput) was inserted instead.  When tracing was enabled, the above code would cause delayedInput to be assigned null as it can no longer be upcast the CDelayedInput.

For each CDelayedInput it was calling setInput (i.e. delayedInput->setInput()) with a new CSplitterOutput() or new CNullInput().  

I modified the above code to 

IThorDataLink *delayedInput = (IThorDataLink *)outputs.item(o);

so that even if the 'outputs' IThorDataLink * items were CTracingThorDataLink *, it could still be assigned to delayedInput.

I added a new setInputSplitterOutputBase() virtual member function , to IThorDataLink, so that 'input' field could be assigned the new CSplitterOutput() or new CNullInput(), even if the  CTracingThorDataLink had wrapped the CDelayedInput object.  It's not ideal but I haven't come up with anything more elegant.

I renamed the original name setInput() to setInputSplitterOutputBase() as setInput was used elsewhere to set the input IThorDataLink but this was setting Owned<CSplitterOutputBase> input - a different type of object.

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -163,7 +163,7 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
 
         CDelayedInput(NSplitterSlaveActivity &_activity) : CThorDataLink(&_activity), activity(&_activity), id(0) { }
-        void setInput(CSplitterOutputBase *_input, unsigned _id=0)
+        virtual void setInputSplitterOutputBase(CSplitterOutputBase *_input, unsigned _id=0)
```
Why has this method name changed?

shamser replied:
Please see previous comment.

jakesmith replied:
Ok, I sort of understand why the change was made now.

Splitter doesn't know how many outputs will be connected up upfront.
It creates pseudo outputs (CDelayInput objects with are IThorDataLink's), where the input can be connected at splitter start() time.

At splitter start time, the existing code assumes the outputs are still instances of what it added (because they were guaranteed to be before) and casts them to CDelayedInput, so as to call the class member setInput().

The new member added to IThorDataLink in this PR (setInputSplitterOutputBase()) , removes the need for casting by introducing it into the common IThorDataLink interface, where all but splitter CDelayedInput implementation do nothing.

However this member isn't something activities in general want to use/know about, so it doesn't belong in IThorDataLink.

I think you achieve the same as the pre-existing code, by tracking the CDelayedInput objects created in the splitter code separately (instead of using 'outputs') and use them directly instead.

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -237,22 +237,22 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
+                    delayedInput->setInputSplitterOutputBase(new CInputWrapper(*this, inputs.item(0)));
                 else
-                    delayedInput->setInput(new CNullInput());
+                    delayedInput->setInputSplitterOutputBase(new CNullInput());
```
I don't understand why these changes were made, can you elaborate?

shamser replied:
Explained couple of comments above.

jakesmith commented on thorlcr/graph/thgraph.cpp:
```diff
@@ -1489,6 +1489,32 @@ void CGraphBase::end()
+        thorDataLinkTracingProbe->remove(edgeId);
+};
+
+
```
Are addThorDebugProbe()/queryThorDataProbe() and removeThorDataProbe() redundant now?
I don't see them used anywhere.

shamser replied:
These will be removed.

jakesmith commented on thorlcr/graph/thgraph.hpp:
```diff
@@ -709,6 +713,10 @@ class graph_decl CGraphBase : public CInterface, implements IEclGraphResults, im
     virtual IEclGraphResults *evaluate(unsigned parentExtractSz, const byte * parentExtract);
+    virtual void addThorDebugProbe(const char *edgeId, IThorDataLink *p);
+    virtual IThorDataLink *queryThorDataProbe(const char *edgeId);
+    virtual void removeThorDataProbe(const char *edgeId);
```
Are addThorDebugProbe()/queryThorDataProbe() and removeThorDataProbe() redundant now?
I don't see them used anywhere.

shamser replied:
Yes.  They'll be removed.

jakesmith commented on thorlcr/graph/thgraphslave.cpp:
```diff
@@ -140,8 +140,29 @@ void CSlaveActivity::setInput(unsigned index, CActivityBase *inputActivity, unsi
     while (inputs.ordinality()<=index) inputs.append(NULL);
-    inputs.replace(outLink.getClear(), index);
+
+    inputs.replace( outLink.getClear(), index);
```
spurious space before 'outLink'

jakesmith commented on thorlcr/graph/thgraphslave.cpp:
```diff
@@ -140,8 +140,29 @@ void CSlaveActivity::setInput(unsigned index, CActivityBase *inputActivity, unsi
+    {
+        const unsigned numTraceRows = queryJob().getOptInt("numTraceRows", 10);
+        Owned<IThorDataLink> newItdl;
+        newItdl.setown(new CTracingThorDataLink(itdl,queryHelper(),numTraceRows));
```
trivial: space after comma

jakesmith replied:
probably no point here in the newItdl declaration, setown, then getClear(), i.e. may as well do:
```
outputs.append(new CTracingThorDataLink(itdl,queryHelper(),numTraceRows));
```

jakesmith commented on thorlcr/graph/thgraphslave.cpp:
```diff
@@ -1151,6 +1177,11 @@ bool CJobSlave::getWorkUnitValueBool(const char *prop, bool defVal) const
 
+void CJobSlave::debugRequest(CMessageBuffer &msg, const char *request) const
+{
+    watchdog->debugRequest(msg, request);
```
wathdog can be disabled, should do a if (watchdog) here and anywhere else similar to this.

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+
+            const char *command = queryXml->queryName();
+            const char *edgeId = queryXml->queryProp("@edgeId");
+            const char *graphId = job->queryGraphName()?job->queryGraphName():"";
```
curious why the :"" ?

shamser replied:
Can queryGraphName return null?

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+            const char *edgeId = queryXml->queryProp("@edgeId");
+            const char *graphId = job->queryGraphName()?job->queryGraphName():"";
+
+            if (!command || !edgeId) throw MakeStringException(5300, "Invalid debug command");
```
shouldn't this also check if !graphId ?

shamser replied:
I can add that check in.  I assume there is no situation when the graph can be null or empty string?

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+            {
+                try
+                {
+                    ISocket *client = sock->accept(true);
```
picky (and debatable): Think looks clearer if this is a Owned<ISocket> and then the CSafeSocket line below is: CSafeSocket ssock(client.getClear());

jakesmith commented on thorlcr/slave/slave.hpp:
```diff
@@ -87,6 +88,8 @@ interface IThorDataLink : extends IRowStream
     virtual void dataLinkSerialize(MemoryBuffer &mb)=0;
     virtual unsigned __int64 queryTotalCycles() const=0;
+    virtual void debugRequest(MemoryBuffer &mb) = 0;
+    virtual void setInputSplitterOutputBase(CSplitterOutputBase *_input, unsigned _id=0) = 0;
```
This doesn't look right.. a method to do with Splitter activity being a virtual member of interface IThorDataLink.

shamser replied:
Please see the long explanation regarding setInputSplitterOutputBase()

jakesmith commented on thorlcr/slave/slavmain.cpp:
```diff
@@ -511,6 +511,25 @@ class CJobListener : public CSimpleInterface
+                            PROGLOG("DebugRequest: %s %s", jobKey.get(), rawText.get());
+                            msg.clear();
+                            job->debugRequest(msg, rawText);
+                        } else 
```
style: normally else on it's own line.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+        {
+            unsigned rowPos = rowTailp + 1; // this is the tail
+            if (rowPos==rowBufferSize) rowPos = 0;
+            // NOTE - we don't output every item in the list as one
```
'as one' -> 'as the last one' ?

shamser replied:
sorry, didn't understand this one.  please can elaborate.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+                if (rowBuffer[rowPos])
+                {
+                  traceOutQueue.enqueue(rowBuffer[rowPos].get(),rowBuffer[rowPos].getSequence());
+                  rowBuffer[rowPos].clear();
```
2 space indentation should be 4.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+                if (++rowPos==rowBufferSize) rowPos = 0; // NOTE - skipping first item intentional
+            }
+        }
+        // Dump items to MemoryBuffer (without clearing items form queue)
```
Typo: 'form' -> 'from'

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+                    xmlwrite.outputInt(recSize, sizeof(recSize), "@size");
+                    try
+                    {
+
```
trivial: spurious newline

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+
+                        meta->toXML((const byte*)(rowBuffer[rowPos].get()), xmlwrite);
+                    }
+                    catch (...)
```
should handle with an IException and trace what exception is.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+        buffers[inactiveBuf].queueOut(rowBufferLogCache);
+
+        atomic_set(&rowBufInUse, inactiveBuf);// Swap Active & inactiveBuf
+        buffers[bufToDump].queueOut(rowBufferLogCache);
```
I think the 2 queueOut calls follow from our previous conversation:
```
Also (and maybe we discussed this(?), it does mean that if there's that race condition, that rows will appear out of sequence. e.g.:

buffers[0] = [a],[b],*c
dump swaps, a,b added to tracing, buffers[1] now active
buffers[1] = [d],[e],*f
dump swap, d,e added to tracing, buffers[0] now active
buffers[0] = [],[],[c],[g],[h],*i
dump swap, c,g,h added to tracing, buffers[1] now active
buffers[1] = [],[],[f],[j],[k],*l
dump swap, f,j,k added to tracing..
Should we just accept that, and just put a comment to mention that this can happen?

Or could we look at the other buffer (the one about to be swapped in) for the trailing row and add to the tracing circular ring, before swapping in ?
I think that could mean there's still a window if 2 swaps happened very quickly, where it would still be active and not picked up and therefore still lead to out order rows, but in practice with the time taken between 2 swaps over 2 debugRequest() hops that isn't going to happen.
```

It would be worth adding a comment to say why callling queueOut for existing (swapped out) buffer, then again for newly swapped out buffer.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,203 @@
+                if (rowBuffer[rowPos])
+                {
+                    xmlwrite.outputBeginNested("Row", true);
+                    xmlwrite.outputInt(rowBuffer[rowPos].getSequence(), sizeof(int), "@seq");
```
It is not clear why CThorRowHistory + getSequence() is good/necessary.
I assume related to previous discussion re. potential for out order ?
But conclusion then was that appending and flushing the already swapped out list + newly swapped out list would take care of that.

shamser replied:
Sequence is provided in HThor and Roxie's debugger response.  I assume that there is a reason for providing this to ECLIDE?

jakesmith commented:
@shamser - back to you. Am a little confused about a couple of changes in particular.

richardkchapman commented:
Also needs rebasing for a clean merge

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

shamser commented:
@richardkchapman @jakesmith Modified as per code review comments & now also returns age of row.  Please let me know if you are ready for me to rebase.

richardkchapman commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+            const char *edgeId = queryXml->queryProp("@edgeId");
+            const char *graphId = job->queryGraphName();
+
+            if (!command || !edgeId || !graphId) throw MakeStringException(5300, "Invalid debug command");
```
graphId being null is not really an "Invalid debug command" scenario, it's more of a "Job cannot be debugged at this time" scenario, isn't it?

jakesmith replied:
The queryGraphName() here could never be null.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,226 @@
+                        StringBuffer msg;
+                        e->errorMessage(msg);
+                        e->Release();
+                        DBGLOG("Row Tracing Error: %s", msg.str());
```
minor: I think exception text and "Row" will be conjoined.
 Normally use EXCLOG(e, <message>); in this scenario.

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+            if (strncmp(command,"print", 5) == 0)
+                cmdcode = DebugRequest;
+            else if (strncmp(command,"quit", 4) == 0)
+                cmdcode = GraphAbort;
```
Is this a good idea?
@richardkchapman  - I'm guessing it's something that Roxie/Hthor supports via debugger?
You can abort the workunit already via the workunit abort button though.

shamser replied:
@jakesmith ECLIDE sends "quit" command when the abort button is pressed.  It also sends other commands such as "continue", "step" etc.  Commands other than "print" and "quite" are ignored.

jakesmith replied:
It (IDE) should know what's supported, i.e. that it's targeting Thor and not enable functions that are not supported.
Or perhaps (in some future version) it could query for list of commands that are supported.

Quitting like this may cause unclean abort sequence (I'm not sure what will happen) - normally a job is aborted either when there's a failure and exceptions trickle up/around, or by CJobBase::abort() call, which e.g. can happen by user request via eclwatch workunit.
If you look for '"ABORT detected from user" in thgraphmaster.cpp, you can see how it handles it.
The debugger could handle "quit" in same way (that would cleaner), i.e. call job->fireException with TE_WorkUnitAborting.

jakesmith replied:
We agreed that this should be changed to call job->abort()

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+
+            const char *command = queryXml->queryName();
+            const char *edgeId = queryXml->queryProp("@edgeId");
+            const char *graphId = job->queryGraphName();
```
Is this right?
graphId vs graphName .. graphName will be something like graph1. A graphId is a numeric that refers to a specific subgraph.

The graphId here isn't being used except as a header in the response. Shouldn't it be passed to the slaves?

@richardkchapman - Does debugger in Roxie/HThor send in a graphId to be debugged?

shamser replied:
@jakesmith The debugger in roxie/hthor returns the graphName in the graphId field.  Not sure how ECLIDE uses it.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,226 @@
+        }
+        inline void enqueue(const CThorRowHistory &_rowBuffer)
+        {
+            enqueue(_rowBuffer.get(),_rowBuffer.getSequence(),_rowBuffer.getmsTickStamp());
```
Still concerned extra complication and usefulness  of sequence and also now timestamp (the cost of it)

jakesmith commented:
Am not convinced or understand the usefulness of sequence and am concerned about cost of sequence and timestamp msTick() for every row on every edge. It's also complicated the implementation.

Do we need sequence per row?
A debug request will 1st queueOut any remaining rows in the inactive queue (which will be oldest) and then all but the potentially active tail from the active queue, from oldest to newest.
I don't think it's possible to be out of sequence, I'm not sure it's possible for there to be a gap in sequence either (?).
If there's a row in the inactive queue being queued (which may be old), then it will only remain in the trace queue if there < max rows in the active queue. Sequence is preserved.
If there are max items in the queue, the trace queue will be filled with all the rows from the active queue that's just been swapped out. Sequence is preserved.

TimeStamp - could be useful sometimes, but msTick() is too expensive to be called per row, per edge.
Could use get_cycles_now(), which in fact every edge already calls (see ActivityTimer), if we are going to keep timestamps, it would be good to call once per nextRow() and reuse it for debugging.
Is it worth it though?

@richardkchapman , I'm not sure how roxie handles this ? It may be less of a concern in Roxie because debugging is more dynamic and on/off, and not something you'd expect to be collecting rows on every edge.

jakesmith commented on thorlcr/activities/thactivityutil.cpp:
```diff
@@ -303,6 +303,8 @@ class ThorLookaheadCache: public IThorDataLink, public CSimpleInterface
     }
     unsigned __int64 queryTotalCycles() const { return in->queryTotalCycles(); }
+    virtual void debugRequest(MemoryBuffer &msg) { return in->debugRequest(msg); }
+    virtual void setInputSplitterOutputBase(CSplitterOutputBase *_input, unsigned _id=0) {}
```
See previous comment re. splitter use of CDelayedInput objects.

jakesmith commented on thorlcr/activities/thactivityutil.ipp:
```diff
@@ -165,7 +165,8 @@ public:
     } 
-
+    virtual void debugRequest(MemoryBuffer &msg) {}
+    virtual void setInputSplitterOutputBase(CSplitterOutputBase *_input, unsigned _id=0) {}
```
See previous comment re. splitter use of CDelayedInput objects.

jakesmith commented on thorlcr/graph/thgraph.hpp:
```diff
@@ -547,6 +550,7 @@ class graph_decl CGraphBase : public CInterface, implements IEclGraphResults, im
     MemoryBuffer parentExtractMb; // retain copy, used if slave transmits to master (child graph 1st time initialization of global graph)
     unsigned counter;
     CReplyCancelHandler graphCancelHandler;
+    Owned<MapStringToIThorDataLink> thorDataLinkTracingProbe;
```
can now be removed.

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+    public:
+        CThorDebugListener(CJobManager &_mgr) : threaded("CThorDebugListener", this), mgr(_mgr)
+        {
+            port = globals->getPropInt("DebugPort", 9877);
```
minor: default should really be in portlist.h, called something like THOR_DEBUG_PORT

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

shamser commented:
@jakesmith Please can you review and let me know when you are ready for me to rebase.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,219 @@
+
+    inline void enqueueRowForTrace(const void *row)
+    {
+        buffers[atomic_read(&rowBufInUse)].enqueue(row,queryEndCycles());
```
The timing here, will actually be the timing of the previous completed nextRow() timing.
The ActivityTimer has called get_cycles_now() again to record local member startCycles, but it is not recorded in the accumulator.

Not sure it really matters that much that recorded time of row is time of last row completion..
Could either comment to explain, or change ActivityTimer so that it stores startCycles in accumulator too and use that.

I think I'd go with the later as easy change and means timing is more honest.

jakesmith replied:
Actually, ignore this point, endCycles is correct.
By the time you are here, you have the row from the real input and it has completed it's ActivityTimer and set endCycles.

jakesmith replied:
minor: you could avoid a virtual call here, by calling thorDataLink->queryEndCycles() here directly, vs virtual queryEndCycles()

jakesmith commented on thorlcr/thorutil/thmem.hpp:
```diff
@@ -88,7 +88,7 @@ class OwnedConstThorRow
     inline OwnedConstThorRow(const OwnedConstThorRow & other)   { ptr = other.getLink(); }
 
-    inline ~OwnedConstThorRow()                             { ReleaseThorRow(ptr); }
+    inline virtual ~OwnedConstThorRow()                     { ReleaseThorRow(ptr); }
```
I am not sure if this is a good idea.. may be ok, but consequence I think is that if you allocate a large number of OwnedConstThorRow objects now, you have a VMT for each where didn't before. So doubles the space required.

@ghalliday - can you comment please?

Not strictly needed as not overloaded by CThorRowHistory.
It may be better if you leave OwnedConstThorRow untouched and contain it as a member of CThorRowHistory

richardkchapman replied:
Almost certainly NOT a good idea

ghalliday replied:
I agree - more correct would be for CThorRowHistory to *contain* a row, not inherit from it.  I'm not convinced it has a is-a relationship.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,219 @@
+    };
+    const unsigned traceQueueSize;
+    class TraceQueue buffers[2];
+    class TraceQueue rowBufferLogCache;
```
don't normally prefix class member definitions with 'class'.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,219 @@
+                    const __int64 age = cycle_to_millisec(cyclesNow-rowCurrent.getCycleStamp());
+                    xmlwrite.outputInt(seq, sizeof(seq), "@seq");
+                    xmlwrite.outputInt(meta->getRecordSize(rowCurrent.get()), sizeof(size32_t), "@size");
+                    xmlwrite.outputInt(age, sizeof(age), "@age");
```
Is @age the format Roxie/HThor output time in? I thought it was a timestamp..

shamser replied:
Roxie/HThor doesn't output the timestamp of the row.  The @age is the amount of time in milliseconds that the row has been in the buffer.  We could just as well provide the timestamp but I would have thought the primary purpose for this is to know how long it is taking to generate new rows.

jakesmith replied:
>Roxie/HThor doesn't output the timestamp of the row.

Ah ok, thought it did.

Timestamp with milliscond definition is more useful, so can tell when a row was output and how long between outputted rows.
There other places we keep duration only and want to convert over to timestamps in fact

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,219 @@
+        {
+            rowBatchCount = seqOfLastRow/rowBufferSize;
+        }
+        inline unsigned getSequence(void) const
```
style: we wouldn't normally declare the void

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on thorlcr/graph/thgraphslave.cpp:
```diff
@@ -140,10 +140,31 @@ void CSlaveActivity::setInput(unsigned index, CActivityBase *inputActivity, unsi
+        const unsigned numTraceRows = queryJob().getOptInt("numTraceRows", 10);
+        Owned<IThorDataLink> newItdl;
+        newItdl.setown(new CTracingThorDataLink(itdl, queryHelper(), numTraceRows));
+        outputs.append(newItdl.getClear());
```
minor: there was a previous comment here, seems to have got lost.
Comment was something along the lines of : Not a lot point in the newItdl here, clearer if above 3 lines replaced with:
```
outputs.append(new CTracingThorDataLink(itdl, queryHelper(), numTraceRows());
```

jakesmith commented:
@shamser - back to you. Main issue left is to remove sequence stuff I think and address a couple of minor comments.

shamser commented:
@jakesmith Done... I'll rebase once I get the go ahead from you.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -317,7 +324,7 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
                     ForEachItemIn(o, outputs)
                     {
-                        CDelayedInput *delayedInput = (CDelayedInput *)outputs.item(o);
+                        IThorDataLink *delayedInput = (IThorDataLink *)outputs.item(o); // TODO: what is delayedInput here for?
```
You're right, delete it.

jakesmith replied:
You're right, worth removing.

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -235,9 +238,9 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
+            ForEachItemIn(o2, delayInputsList)
             {
-                CDelayedInput *delayedInput = (CDelayedInput *)outputs.item(o2);
+                CDelayedInput *delayedInput = (CDelayedInput *)delayInputsList.item(o2);
```
minor: unnecessary cast. (Thought I commented on this before.)

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -246,9 +249,9 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
+            ForEachItemIn(o, delayInputsList)
             {
-                CDelayedInput *delayedInput = (CDelayedInput *)outputs.item(o);
+                CDelayedInput *delayedInput = (CDelayedInput *)delayInputsList.item(o);
```
minor: unnecessary cast. (Thought I commented on this before.)

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -268,9 +271,9 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
+            ForEachItemIn(o, delayInputsList)
             {
-                CDelayedInput *delayedInput = (CDelayedInput *)outputs.item(o);
+                CDelayedInput *delayedInput = (CDelayedInput *)delayInputsList.item(o);
```
minor: unnecessary cast. (Thought I commented on this before.)

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -411,7 +418,7 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
         ForEachItemIn(o, outputs)
         {
-            CDelayedInput *delayedInput = (CDelayedInput *)outputs.item(o);
+            IThorDataLink *delayedInput = (IThorDataLink *)outputs.item(o);
```
minor: unnecessary cast.

jakesmith commented on thorlcr/graph/thgraph.hpp:
```diff
@@ -415,6 +416,8 @@ class graph_decl CGraphTempHandler : public CInterface, implements IGraphTempHan
 class CJobBase;
 interface IPropertyTree;
+class IThorDataLink;
+typedef MapStringTo<IThorDataLink *> MapStringToIThorDataLink;
```
minor: no longer used.

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,163 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+            else if (strncmp(command,"quit", 4) == 0)
+            {
+                LOG(MCwarning, thorJob, "ABORT detected from user");
+                Owned <IException> e = MakeThorException(TE_WorkUnitAborting, "User signalled abort");
```
quibble: don't normally have space between Owned and <

shamser replied:
Note, should I also modify the original code where there is a space there also?  (thgraphmaster.cpp line 1637)

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,206 @@
+        buffers[0].init(traceQueueSize+1);
+    }
+
+    virtual ~CTracingThorDataLink() { }
```
not needed.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

shamser commented:
@jakesmith Modified as per code review comments.

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,227 @@
+
+            const __int64 diffUSec = cycle_to_microsec(get_cycles_now() - cycleStamp);
+            const __int64 adjustSecs = diffUSec / 1000000;
+            const __int64 adjustUSecs = (diffUSec % 1000000 );
```
trivial: extra space before )

jakesmith commented on thorlcr/slave/traceslave.hpp:
```diff
@@ -0,0 +1,227 @@
+            {
+                tv.tv_sec -= (adjustSecs+1);
+                tv.tv_usec = tv.tv_usec + 1000000 - adjustUSecs;
+            };
```
Did you look at using timersub here instead?

jakesmith commented on thorlcr/activities/nsplitter/thnsplitterslave.cpp:
```diff
@@ -163,7 +163,7 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
 
         CDelayedInput(NSplitterSlaveActivity &_activity) : CThorDataLink(&_activity), activity(&_activity), id(0) { }
-        void setInput(CSplitterOutputBase *_input, unsigned _id=0)
+        virtual void setInput(CSplitterOutputBase *_input, unsigned _id=0)
```
lost comment I think, doesn't need to be virtual

jakesmith commented:
@shamser - final comment I think, will also nee rebasing.

shamser commented:
@jakesmith @richardkchapman Commit re-based.  Over to you.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@shamser - Looks good, but needs rebasing onto master

shamser commented:
@jakesmith @richardkchapman Rebased to master

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -60,10 +61,163 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+                sock->cancel_accept();
+            threaded.join();
+        }
+        virtual unsigned getPort() { return port; }
```
trivial: could be const

jakesmith commented:
@richardkchapman - ready to merge

shamser committed:
HPCC-13030 Add row tracing support to engines

CDelayedInput object pointer stored in multiple locations - usage must be tracked

shamser commented:
@jakesmith bug fix: was storing CDelayedInput in new array but not tracking usage.

shamser committed:
HPCC-13030 Add row tracing support to engines

Make .getPort() const

HPCCSmoketest commented:
Automated Smoketest Build: failed
Number of Error(s): 0

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

GordonSmith commented on thorlcr/graph/thgraphslave.cpp:
```diff
@@ -140,10 +140,29 @@ void CSlaveActivity::setInput(unsigned index, CActivityBase *inputActivity, unsi
+    if (queryJob().getOptBool("TRACEROWS"))
+    {
+        const unsigned numTraceRows = queryJob().getOptInt("numTraceRows", 10);
+        outputs.append(new CTracingThorDataLink(itdl, queryHelper(), numTraceRows));
```
Will this work if an output ever gets cast (or query interfaced) back to its original base class?

What I am seeing is that the metaInfo (counts and recordSize etc.) is not getting reported within the graphs and suspect that something casting *CTracingThorDataLink to *CThorDataLink (via queryOutput?) and if successful pulls the metaInfo from it?

shamser replied:
No, if outputs is cast back to the original base class, it won't work.  We had to make changes to NSplitterSlaveActivity because of this issue.  Let me see everywhere that outputs and, if necessary, modify to work with CTracingThorDataLink.

jakesmith replied:
I don't think there is (or was) any casting to the implemented class, except NSplitter's special case where it was casting to it's own derivative of CThorDataLink, which as @shamser says has been handled differently now (not using queryOutput()).

If there were any miscasting, you'd likely get a crash than missing meta info.

@GordonSmith, is the missing meta info involving particular activities/edges ?

shamser replied:
@GordonSmith The missing meta information is occuring in the master branch.   All work-units (even one produced by prior builds) has missing metat information. If I revert to the 
candidate-5.4.6, the meta information is visible again.  A problem with eclwatch? Jira ticket: https://track.hpccsystems.com/browse/HPCC-14600

shamser committed:
HPCC-13030 Add row tracing support to engines

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
Merged manually to 6.0.0, squashed to single commit
